content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst

```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
-We need a way to compose an Address value from an integer value. The 
-package :ada:`System.Storage_Elements` defines an integer representation 
-for Address values, among other useful things, and a way to convert 
-those integer values to Address values. The function :ada:`To_Address` 
-does that conversion. 
+We declare the object and also specify the address, but not by querying
+some entity. We already know the address from the hardware
+documentation. But we cannot simply use an integer address literal from
+that documentation because type :ada:`System.Address` is almost always a
+private type. We need a way to compose an :ada:`Address` value from an
+integer value. The package :ada:`System.Storage_Elements` defines an
+integer representation for :ada:`Address` values, among other useful
+things, and a way to convert those integer values to :ada:`Address`
+values. The function :ada:`To_Address` does that conversion.
+
+As a result, in the Ada code, reading the value of the variable
+Rotary_Switch reads the number on the actual hardware switch.
+
+Note that if you specify the wrong address, it is hard to say what
+happens. Likewise, it is an error for an address clause to disobey the
+object's alignment. The error cannot be detected at compile time, in
+general, because the address is not necessarily known at compile time.
+There's no requirement for a run-time check for the sake of efficiency,
+since efficiency seems paramount here. Consequently, this misuse of
+address clauses is just like any other misuse of address clauses |mdash|
+execution of the code is erroneous, meaning all bets are off. You need
+to know what you're doing.
+
+What about writing to the variable? Is that meaningful? In this
+particular rotary switch case, no. It is effectively read-only memory.
+But for some other device it very well could be meaningful, certainly.
+It depends on the hardware. But in this case, assigning a value to the
+Rotary_Switch variable would have no effect, which could be confusing to
+programmers. It looks like a variable, after all. We wouldn't declare it
+as a constant because the human user could rotate the switch, resulting
+in a different value read. Therefore, we would hid the Ada variable
+behind a function, obviating the entire question. Clients of the
+function can then use it for whatever purpose they require, e.g., as the
+unique identifier for a computer in a rack.
+
+Let's talk more about the type we use to represent a memory-mapped
+device. As we said, that type defines the view we have for the object,
+and hence the operations we have available for accessing the underlying
+mapped device.
+
+We choose the type for the representative Ada variable based on the
+interface of the hardware mapped to the memory. If the interface is a
+single monolithic register, for example, then an integer, either signed
+or unsigned, and of the necessary size, will suffice. But suppose the
```

Comment:
This phrase sounds a bit confusing to me. (Maybe there are too many commas here.) As a suggestion, what about using an m-dash?

```suggestion
interface of the hardware mapped to the memory. If the interface is a
single monolithic register, for example, then an integer |mdash| either signed
or unsigned, and of the necessary size |mdash| will suffice. But suppose the
```

Reply:
Or maybe make it parenthetical, like this:

If the interface is a single monolithic register, for example, then an integer (signed
or unsigned) of the necessary size will suffice.

Reply:
That sounds good to me.